core/remotes/docker/fetcher.go

```diff
@@ -262,74 +437,122 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
+	switch err {
+	case nil:
+		// all good
+	case errContentRangeIgnored:
+		// remote host ignored content range, force parallelism to 1
+		parallelism = 1
+		err = nil
+	default:
+		log.G(ctx).WithError(err).Debug("fetch failed")
+		r.Release(1)
 		return nil, err
 	}
-	defer func() {
-		if retErr != nil {
-			resp.Body.Close()
-		}
-	}()
 
-	if resp.StatusCode > 299 {
-		// TODO(stevvooe): When doing a offset specific request, we should
-		// really distinguish between a 206 and a 200. In the case of 200, we
-		// can discard the bytes, hiding the seek behavior from the
-		// implementation.
+	body := resp.Body
+	encoding := strings.FieldsFunc(resp.Header.Get("Content-Encoding"), func(r rune) bool {
+		return r == ' ' || r == '\t' || r == ','
+	})
 
-		if resp.StatusCode == http.StatusNotFound {
-			return nil, fmt.Errorf("content at %v not found: %w", req.String(), errdefs.ErrNotFound)
+	remaining, _ := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	if parallelism > 1 && chunkSize > 512 && req.body == nil {
+		// If we have a content length, we can use multiple requests to fetch
+		// the content in parallel. This will make download of bigger bodies
+		// faster, at the cost of parallelism more requests and max
+		// ~(max_parallelism * goroutine footprint) memory usage. The goroutine
+		// footprint should be: the goroutine stack + pipe buffer size
+		numChunks := remaining / chunkSize
+		if numChunks*chunkSize < remaining {
+			numChunks++
 		}
-		var registryErr Errors
-		if err := json.NewDecoder(resp.Body).Decode(&registryErr); err != nil || registryErr.Len() < 1 {
-			return nil, fmt.Errorf("unexpected status code %v: %v", req.String(), resp.Status)
+		if numChunks < parallelism {
+			parallelism = numChunks
 		}
-		return nil, fmt.Errorf("unexpected status code %v: %s - Server message: %s", req.String(), resp.Status, registryErr.Error())
-	}
-	if offset > 0 {
-		cr := resp.Header.Get("content-range")
-		if cr != "" {
-			if !strings.HasPrefix(cr, fmt.Sprintf("bytes %d-", offset)) {
-				return nil, fmt.Errorf("unhandled content range in response: %v", cr)
-
-			}
-		} else {
-			// TODO: Should any cases where use of content range
-			// without the proper header be considered?
-			// 206 responses?
-
-			// Discard up to offset
-			// Could use buffer pool here but this case should be rare
-			n, err := io.Copy(io.Discard, io.LimitReader(resp.Body, offset))
-			if err != nil {
-				return nil, fmt.Errorf("failed to discard to offset: %w", err)
-			}
-			if n != offset {
-				return nil, errors.New("unable to discard to offset")
+		queue := make(chan int64, parallelism)
+		stopChan := make(chan struct{})
+		readers, writers := make([]io.Reader, numChunks), make([]*pipeWriter, numChunks)
+		bufPool := newbufferPool(chunkSize)
+		for i := range numChunks {
+			readers[i], writers[i] = newPipeWriter(bufPool)
+		}
+		go func() {
+			for i := range numChunks {
+				queue <- i
```

Comment:
Should this also select on `stopChan` to prevent leaking of this go routine? The queue will stop draining once an error has occurred

Reply:
Good point, adding !

Reply:
done and tested a bunch, lmk if anything !